【WIP】新しいソースを追加した際に自動でキャンバスに収まる最大サイズまで拡縮するように変更

takayamaki opened the pull request:
**このpull requestが解決する内容**
新しいソースを追加した際に自動でキャンバスに収まる最大サイズまで拡縮します。

- テキストソース、カラーソース、音声入力キャプチャ、音声出力キャプチャでは発動しません
  - テキストソースはfitToScreenするとフォントサイズは変わらないまま画像処理で拡大されるので解像度不足でボケてしまうため
  - カラーソースはソースのプロパティに縦横サイズが存在するため
  - 音声入力/出力キャプチャはfitToScreenする意味がないため
- 追加直後かつキャンセルボタンが押下されたときに発動させるかどうか迷いましたが、とりあえずさせる方向にしました

![2018-12-05 11-51-48](https://user-images.githubusercontent.com/24884114/49487319-2e8afe80-f885-11e8-884e-babcf9c714ef.gif)

related #210
**動作確認手順**

takayamaki committed:
新しいソースを追加した際に自動でキャンバスに収まる最大サイズまで拡縮するように変更

close #210

berlysia commented:
ある拡縮対象種のソース（たとえば画像ソース）がシーン中に存在しない状態からだとフィットしますが、すでに存在しているとフィットしないです。概要のgifの状態からもうひとつ画像ソースを足してみてください

ここかな
https://github.com/n-air-app/n-air-app/blob/11ddfe061a355c1cf8654cc0ab3e5eeb7c873245/app/components/windows/AddSource.vue.ts#L83

takayamaki committed:
AddSource.vue.tsへの改修が漏れていたため修正

- 同じ種類のソースの2個目以降を追加する際に
  - 新規ソースとして追加するときにはshowSourcePropertiesの第二引数にtrueを渡して設定完了後にfitToScreenする
  - 複製を追加するときにはshowSourcePropertiesを経由しないので、すぐにfitToScreenする

takayamaki commented:
コメントを受けて修正しました。

参照コピー追加の際も、ソースの設定は引き継がれるものの配置は拡大率100%の座標0,0に戻る仕様だったため、ISceneApiのaddSourceの戻り値がISceneItemApiだったので即fitToScreenするようにしました。

berlysia commented on app/components/windows/AddSource.vue.ts:
```diff
@@ -53,7 +53,8 @@ export default class AddSource extends Vue {
     }
-    this.scenesService.activeScene.addSource(this.selectedSourceId);
+    const addedItem = this.scenesService.activeScene.addSource(this.selectedSourceId);
+    addedItem.fitToScreen();
```
種類の判定をしていないので、テキストソースで既存のソースを追加した場合にfitToScreenが呼ばれました

takayamaki committed:
適用対象ではないはずのソースタイプにまでfitToScreenしてしまっていた問題を修正

- isRequireFitToScreenをAutoFitToScreen.isRequiredとして切り出しimportするよう変更
  - AddSource.vue.tsにも書く必要があるが、重複ロジック化を防ぐため

takayamaki commented:
重複ロジックとなることを嫌って別ファイルに切り出しましたが、切り出し方がこれでよいのかどうかは自信がないのでレビューお願いできればと思います。

berlysia commented:
色々考えることがありそうに思います。

## ソースタイプに対応して自動的に fitToScreen を呼ぶことについて

大きいソースが与えられた時に小さくするのが煩雑、というのは同意します

しかし、ソースタイプで分類して fitToScreen を呼んでしまうと、十分に小さくて取り回しやすい状態で生成されたシーンアイテムが拡大されてしまい、また別の煩雑さを生み出すと思います。

対策するとしたら、 fitToScreen のように呼び出せて、スクリーンより大きいときにはスクリーンサイズまで小さくしてくれて、必要なければ何もしないようなメソッドを追加し、呼び出すことが考えられるでしょうか

例えばこれの近くに並べられそう
https://github.com/n-air-app/n-air-app/blob/n-air_development/app/services/scenes/scene-item.ts#L286-L290

## 新規作成ソースの場合の動作について

ソースプロパティの画面を完了/キャンセルで閉じたときに発動するのは、内部実装の都合上こんな感じになるしかないかなと思っています。

利用者が驚きそうなパターンとして、新規追加時のソースプロパティ画面を無視して main ウィンドウの操作を始めてしまい、思い出した頃にソースプロパティ画面を閉じたところ fitToScreen が呼ばれる、というシナリオが考えられそうです。
ソースプロパティ画面（をはじめとする既存の child ウィンドウを使う画面）は現在全てモードレスなので、画面を無視して main ウィンドウを操作できますよね。

なのでこの形をそのまま活かそうとするなら、child ウィンドウをモーダルとして使えるように拡張する必要があるかと思います。少なくともchildウィンドウを操作しなければならないというアフォーダンスが欲しいです。

---

関連して、拡縮対象のソースを新規作成して、ソースプロパティ画面が開いた状態でシーンを変更し、ソースプロパティ画面を閉じると fitToScreen が呼ばれないようなのでご確認ください。
